File: alica_engine/src/engine/modelmanagement/factories/ConditionFactory.cpp
#include "engine/modelmanagement/factories/ConditionFactory.h"
#include "engine/model/AbstractPlan.h"
#include "engine/model/Condition.h"
#include "engine/modelmanagement/Strings.h"
#include "engine/modelmanagement/factories/QuantifierFactory.h"

namespace alica
{
void ConditionFactory::fillCondition(const YAML::Node& conditionNode, Condition* condition, alica::AbstractPlan* abstractPlan)
{
    Factory::setAttributes(conditionNode, condition);
    Factory::storeElement(condition, alica::Strings::condition);
    condition->setAbstractPlan(abstractPlan);
    condition->setConditionString(Factory::getValue<std::string>(conditionNode, alica::Strings::conditionString, ""));

    if (Factory::isValid(conditionNode[alica::Strings::libraryName])) {
        condition->setLibraryName(Factory::getValue<std::string>(conditionNode, alica::Strings::libraryName));
    } else {
        condition->setLibraryName(static_cast<Plan*>(abstractPlan)->getLibraryName());
    }

    if (Factory::isValid(conditionNode[alica::Strings::variables])) {
        const YAML::Node& variableNodes = conditionNode[alica::Strings::variables];
        for (YAML::const_iterator it = variableNodes.begin(); it != variableNodes.end(); ++it) {
            Factory::conditionVarReferences.push_back(std::pair<int64_t, int64_t>(condition->getId(), Factory::getReferencedId(*it)));
        }
    }
    if (Factory::isValid(conditionNode[alica::Strings::quantifiers])) {
        const YAML::Node& quantifierNodes = conditionNode[alica::Strings::quantifiers];
        for (YAML::const_iterator it = quantifierNodes.begin(); it != quantifierNodes.end(); ++it) {
            condition->addQuantifier(QuantifierFactory::create(*it));
        }
    }
}

void ConditionFactory::attachReferences()
{
    QuantifierFactory::attachReferences();

    for (std::pair<int64_t, int64_t> pairs : Factory::conditionVarReferences) {
        Condition* c = (Condition*) Factory::getElement(pairs.first);
        Variable* v = (Variable*) Factory::getElement(pairs.second);
        c->addVariable(v);
    }
    Factory::conditionVarReferences.clear();
}
} // namespace alica

File: alica-supplementary/plan-conversion-tool/src/main/java/com/rapyutarobotics/alica/factories/ConditionFactory.java
package com.rapyutarobotics.alica.factories;

import com.rapyutarobotics.alica.ConversionProcess;
import com.rapyutarobotics.alica.Tags;
import de.unikassel.vs.alica.planDesigner.alicamodel.Condition;
import de.unikassel.vs.alica.planDesigner.alicamodel.Variable;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;

import java.util.HashMap;

public class ConditionFactory extends Factory {
    public static void fillConditon(Element conditionNode, Condition condition, ConversionProcess cp) {
        Factory.setAttributes(conditionNode, condition);
        cp.addElement(condition);
        condition.setConditionString(conditionNode.getAttribute(Tags.CONDITIONSTRING));
        condition.setPluginName(conditionNode.getAttribute(Tags.PLUGINNAME));
        NodeList varNodes = conditionNode.getElementsByTagName(Tags.VARIABLES);
        for (int i = 0; i  < varNodes.getLength(); i++) {
            Element varNode = (Element) varNodes.item(i);
            cp.conditionVarReferences.put(condition.getId(), cp.getReferencedId(varNode.getTextContent()));
        }
        NodeList quantifierNodes = conditionNode.getElementsByTagName(Tags.QUANTIFIERS);
        for (int i = 0; i  < quantifierNodes.getLength(); i++) {
            Element quantifierNode = (Element) quantifierNodes.item(i);
            condition.addQuantifier(QuantifierFactory.create(quantifierNode, cp));
        }
    }

    public static void attachReferences(ConversionProcess cp) {
        QuantifierFactory.attachReferences(cp);

        for (HashMap.Entry<Long, Long> entry : cp.conditionVarReferences.entrySet()) {
            Condition condition = (Condition) cp.getElement(entry.getKey());
            Variable variable = (Variable) cp.getElement(entry.getValue());
            condition.addVariable(variable);
        }
        cp.conditionVarReferences.clear();
    }
}
